Unexpected EOF while trying to create DNS record

### Terraform Version

Terraform v0.11.1

### Affected Resource(s)

- dns_a_record_set

### Terraform Configuration Files

Note: names and addresses have been changed in the code snippet below 
The sku of the machine I'm trying to connect is 2012-R2-Datacenter-smalldisk

```hcl
provider "dns" {
  update {
    server = "10.10.10.10"
  }
}

resource "dns_a_record_set" "test" {
  zone = "example.com."
  name = "www."
  addresses = [
    "10.10.10.11",
  ]
}
```

### Debug Output

```
2018/01/09 10:48:15 [DEBUG] apply: dns_a_record_set.test: executing Apply
2018-01-09T10:48:15.285Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: panic: runtime error: invalid memory address or nil pointer dereference
2018-01-09T10:48:15.285Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x9dd9c3]
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: goroutine 55 [running]:
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/miekg/dns.(*Msg).Insert(0xc4202d7850, 0xc4202d7718, 0x1, 0x1)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/miekg/dns/update.go:71 +0xa3
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/dns.resourceDnsARecordSetUpdate(0xc420231960, 0xa9e020, 0xc420071360, 0x2, 0x2)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/dns/resource_dns_a_record_set.go:144 +0x89e
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/dns.resourceDnsARecordSetCreate(0xc420231960, 0xa9e020, 0xc420071360, 0x24, 0x1051b40)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/dns/resource_dns_a_record_set.go:58 +0x2f7
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc4202a33e0, 0xc4201110e0, 0xc42034a1a0, 0xa9e020, 0xc420071360, 0x1, 0xc42034c030, 0xc420280db0)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:193 +0x3b6
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc420231500, 0xc420111090, 0xc4201110e0, 0xc42034a1a0, 0x7f93826c06c8, 0x0, 0x18)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:259 +0xa4
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc4202e9960, 0xc4202e9d80, 0xc420320b10, 0x0, 0x0)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:488 +0x57
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: reflect.Value.call(0xc420130f00, 0xc420118be0, 0x13, 0xbc8b80, 0x4, 0xc420041f20, 0x3, 0x3, 0xc4200266e8, 0xc4200829c8, ...)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/reflect/value.go:434 +0x906
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: reflect.Value.Call(0xc420130f00, 0xc420118be0, 0x13, 0xc420026720, 0x3, 0x3, 0xc400000005, 0x0, 0x0)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: net/rpc.(*service).call(0xc42031c540, 0xc420110ff0, 0xc42011bb78, 0xc420129e80, 0xc4203232c0, 0xaa02a0, 0xc4202e9d80, 0x16, 0xaa02e0, 0xc420320b10, ...)
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: created by net/rpc.(*Server).ServeCodec
2018-01-09T10:48:15.286Z [DEBUG] plugin.terraform-provider-dns_v1.0.0_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2018-01-09T10:48:15.286Z [DEBUG] plugin: plugin process exited: path=/home/qurrent/azureinfrastructure/azureinfrastructure/global/terraform-modules/test/.terraform/plugins/linux_amd64/terraform-provider-dns_v1.0.0_x4
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalWriteState
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalIf
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalWriteState
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalWriteDiff
2018/01/09 10:48:15 [TRACE] root: eval: *terraform.EvalApplyPost
2018/01/09 10:48:15 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* dns_a_record_set.test: unexpected EOF
2018/01/09 10:48:15 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* dns_a_record_set.test: unexpected EOF
2018/01/09 10:48:15 [TRACE] [walkApply] Exiting eval tree: dns_a_record_set.test

```
full output at:
https://gist.github.com/SvenSchipper/1b247f0cf1d1bf7a961453a24c2a344d

### Panic Output

full output at:
https://gist.github.com/SvenSchipper/0ffaab89f4386d8bc6e27a442a3e8fe8

### Expected Behavior

An A type DNS record set will be created.

### Actual Behavior
```
err: 1 error(s) occurred:

* dns_a_record_set.test: unexpected EOF
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids
This is run in the Azure environment

--- a/dns/config.go
+++ b/dns/config.go
@@ -2,10 +2,12 @@ package dns
 
 import (
 	"fmt"
-	"github.com/miekg/dns"
 	"log"
 	"net"
 	"strconv"
+	"strings"
+
+	"github.com/miekg/dns"
 )
 
 type Config struct {
@@ -40,14 +42,18 @@ func (c *Config) Client() (interface{}, error) {
 	}
 	client.c = new(dns.Client)
 	if c.keyname != "" {
-		client.keyname = c.keyname
+		if !dns.IsFqdn(c.keyname) {
+			return nil, fmt.Errorf("Error configuring provider: \"key_name\" should be fully-qualified")
+		}
+		keyname := strings.ToLower(c.keyname)
+		client.keyname = keyname
 		client.keysecret = c.keysecret
 		keyalgo, err := convertHMACAlgorithm(c.keyalgo)
 		if err != nil {
 			return nil, fmt.Errorf("Error configuring provider: %s", err)
 		}
 		client.keyalgo = keyalgo
-		client.c.TsigSecret = map[string]string{c.keyname: c.keysecret}
+		client.c.TsigSecret = map[string]string{keyname: c.keysecret}
 	}
 	return &client, nil
 }

--- a/dns/resource_dns_a_record_set.go
+++ b/dns/resource_dns_a_record_set.go
@@ -17,14 +17,16 @@ func resourceDnsARecordSet() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"zone": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateZone,
 			},
 			"name": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateName,
 			},
 			"addresses": &schema.Schema{
 				Type:     schema.TypeSet,
@@ -47,10 +49,6 @@ func resourceDnsARecordSetCreate(d *schema.ResourceData, meta interface{}) error
 	rec_name := d.Get("name").(string)
 	rec_zone := d.Get("zone").(string)
 
-	if rec_zone != dns.Fqdn(rec_zone) {
-		return fmt.Errorf("Error creating DNS record: \"zone\" should be an FQDN")
-	}
-
 	rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 	d.SetId(rec_fqdn)
@@ -64,10 +62,6 @@ func resourceDnsARecordSetRead(d *schema.ResourceData, meta interface{}) error {
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -130,10 +124,6 @@ func resourceDnsARecordSetUpdate(d *schema.ResourceData, meta interface{}) error
 		rec_zone := d.Get("zone").(string)
 		ttl := d.Get("ttl").(int)
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -184,10 +174,6 @@ func resourceDnsARecordSetDelete(d *schema.ResourceData, meta interface{}) error
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 

--- a/dns/resource_dns_a_record_set_test.go
+++ b/dns/resource_dns_a_record_set_test.go
@@ -76,10 +76,6 @@ func testAccCheckDnsARecordSetDestroy(s *terraform.State) error {
 		rec_name := rs.Primary.Attributes["name"]
 		rec_zone := rs.Primary.Attributes["zone"]
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -108,10 +104,6 @@ func testAccCheckDnsARecordSetExists(t *testing.T, n string, addr []interface{},
 
 		*rec_name = rs.Primary.Attributes["name"]
 		*rec_zone = rs.Primary.Attributes["zone"]
-
-		if *rec_zone != dns.Fqdn(*rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", *rec_name, *rec_zone)
 

--- a/dns/resource_dns_aaaa_record_set.go
+++ b/dns/resource_dns_aaaa_record_set.go
@@ -17,14 +17,16 @@ func resourceDnsAAAARecordSet() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"zone": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateZone,
 			},
 			"name": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateName,
 			},
 			"addresses": &schema.Schema{
 				Type:     schema.TypeSet,
@@ -47,10 +49,6 @@ func resourceDnsAAAARecordSetCreate(d *schema.ResourceData, meta interface{}) er
 	rec_name := d.Get("name").(string)
 	rec_zone := d.Get("zone").(string)
 
-	if rec_zone != dns.Fqdn(rec_zone) {
-		return fmt.Errorf("Error creating DNS record: \"zone\" should be an FQDN")
-	}
-
 	rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 	d.SetId(rec_fqdn)
@@ -64,10 +62,6 @@ func resourceDnsAAAARecordSetRead(d *schema.ResourceData, meta interface{}) erro
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -130,10 +124,6 @@ func resourceDnsAAAARecordSetUpdate(d *schema.ResourceData, meta interface{}) er
 		rec_zone := d.Get("zone").(string)
 		ttl := d.Get("ttl").(int)
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -184,10 +174,6 @@ func resourceDnsAAAARecordSetDelete(d *schema.ResourceData, meta interface{}) er
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 

--- a/dns/resource_dns_aaaa_record_set_test.go
+++ b/dns/resource_dns_aaaa_record_set_test.go
@@ -83,10 +83,6 @@ func testAccCheckDnsAAAARecordSetDestroy(s *terraform.State) error {
 		rec_name := rs.Primary.Attributes["name"]
 		rec_zone := rs.Primary.Attributes["zone"]
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -115,10 +111,6 @@ func testAccCheckDnsAAAARecordSetExists(t *testing.T, n string, addr []interface
 
 		*rec_name = rs.Primary.Attributes["name"]
 		*rec_zone = rs.Primary.Attributes["zone"]
-
-		if *rec_zone != dns.Fqdn(*rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", *rec_name, *rec_zone)
 

--- a/dns/resource_dns_cname_record.go
+++ b/dns/resource_dns_cname_record.go
@@ -16,18 +16,21 @@ func resourceDnsCnameRecord() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"zone": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateZone,
 			},
 			"name": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateName,
 			},
 			"cname": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateZone,
 			},
 			"ttl": &schema.Schema{
 				Type:     schema.TypeInt,
@@ -43,15 +46,6 @@ func resourceDnsCnameRecordCreate(d *schema.ResourceData, meta interface{}) erro
 
 	rec_name := d.Get("name").(string)
 	rec_zone := d.Get("zone").(string)
-	rec_cname := d.Get("cname").(string)
-
-	if rec_zone != dns.Fqdn(rec_zone) {
-		return fmt.Errorf("Error creating DNS record: \"zone\" should be an FQDN")
-	}
-
-	if rec_cname != dns.Fqdn(rec_cname) {
-		return fmt.Errorf("Error creating DNS record: \"cname\" should be an FQDN")
-	}
 
 	rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -67,14 +61,6 @@ func resourceDnsCnameRecordRead(d *schema.ResourceData, meta interface{}) error 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
 		rec_cname := d.Get("cname").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
-		if rec_cname != dns.Fqdn(rec_cname) {
-			return fmt.Errorf("Error reading DNS record: \"cname\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -119,16 +105,7 @@ func resourceDnsCnameRecordUpdate(d *schema.ResourceData, meta interface{}) erro
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-		rec_cname := d.Get("cname").(string)
 		ttl := d.Get("ttl").(int)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
-
-		if rec_cname != dns.Fqdn(rec_cname) {
-			return fmt.Errorf("Error updating DNS record: \"cname\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -174,10 +151,6 @@ func resourceDnsCnameRecordDelete(d *schema.ResourceData, meta interface{}) erro
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 

--- a/dns/resource_dns_cname_record_test.go
+++ b/dns/resource_dns_cname_record_test.go
@@ -72,10 +72,6 @@ func testAccCheckDnsCnameRecordDestroy(s *terraform.State) error {
 		rec_name := rs.Primary.Attributes["name"]
 		rec_zone := rs.Primary.Attributes["zone"]
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -104,10 +100,6 @@ func testAccCheckDnsCnameRecordExists(t *testing.T, n string, expected string, r
 
 		*rec_name = rs.Primary.Attributes["name"]
 		*rec_zone = rs.Primary.Attributes["zone"]
-
-		if *rec_zone != dns.Fqdn(*rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", *rec_name, *rec_zone)
 

--- a/dns/resource_dns_ns_record_set.go
+++ b/dns/resource_dns_ns_record_set.go
@@ -16,20 +16,25 @@ func resourceDnsNSRecordSet() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"zone": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateZone,
 			},
 			"name": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateName,
 			},
 			"nameservers": &schema.Schema{
 				Type:     schema.TypeSet,
 				Required: true,
-				Elem:     &schema.Schema{Type: schema.TypeString},
-				Set:      schema.HashString,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validateZone,
+				},
+				Set: schema.HashString,
 			},
 			"ttl": &schema.Schema{
 				Type:     schema.TypeInt,
@@ -46,10 +51,6 @@ func resourceDnsNSRecordSetCreate(d *schema.ResourceData, meta interface{}) erro
 	rec_name := d.Get("name").(string)
 	rec_zone := d.Get("zone").(string)
 
-	if rec_zone != dns.Fqdn(rec_zone) {
-		return fmt.Errorf("Error creating DNS record: \"zone\" should be an FQDN")
-	}
-
 	rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 	d.SetId(rec_fqdn)
@@ -63,10 +64,6 @@ func resourceDnsNSRecordSetRead(d *schema.ResourceData, meta interface{}) error 
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -116,10 +113,6 @@ func resourceDnsNSRecordSetUpdate(d *schema.ResourceData, meta interface{}) erro
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
 		ttl := d.Get("ttl").(int)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -171,10 +164,6 @@ func resourceDnsNSRecordSetDelete(d *schema.ResourceData, meta interface{}) erro
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 

--- a/dns/resource_dns_ns_record_set_test.go
+++ b/dns/resource_dns_ns_record_set_test.go
@@ -76,10 +76,6 @@ func testAccCheckDnsNSRecordSetDestroy(s *terraform.State) error {
 		rec_name := rs.Primary.Attributes["name"]
 		rec_zone := rs.Primary.Attributes["zone"]
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -109,10 +105,6 @@ func testAccCheckDnsNSRecordSetExists(t *testing.T, n string, nameserver []inter
 
 		*rec_name = rs.Primary.Attributes["name"]
 		*rec_zone = rs.Primary.Attributes["zone"]
-
-		if *rec_zone != dns.Fqdn(*rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", *rec_name, *rec_zone)
 

--- a/dns/resource_dns_ptr_record.go
+++ b/dns/resource_dns_ptr_record.go
@@ -16,18 +16,21 @@ func resourceDnsPtrRecord() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"zone": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateZone,
 			},
 			"name": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateName,
 			},
 			"ptr": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateZone,
 			},
 			"ttl": &schema.Schema{
 				Type:     schema.TypeInt,
@@ -43,15 +46,6 @@ func resourceDnsPtrRecordCreate(d *schema.ResourceData, meta interface{}) error 
 
 	rec_name := d.Get("name").(string)
 	rec_zone := d.Get("zone").(string)
-	rec_ptr := d.Get("ptr").(string)
-
-	if rec_zone != dns.Fqdn(rec_zone) {
-		return fmt.Errorf("Error creating DNS record: \"zone\" should be an FQDN")
-	}
-
-	if rec_ptr != dns.Fqdn(rec_ptr) {
-		return fmt.Errorf("Error creating DNS record: \"ptr\" should be an FQDN")
-	}
 
 	rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -67,14 +61,6 @@ func resourceDnsPtrRecordRead(d *schema.ResourceData, meta interface{}) error {
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
 		rec_ptr := d.Get("ptr").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
-		if rec_ptr != dns.Fqdn(rec_ptr) {
-			return fmt.Errorf("Error reading DNS record: \"ptr\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -119,16 +105,7 @@ func resourceDnsPtrRecordUpdate(d *schema.ResourceData, meta interface{}) error 
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-		rec_ptr := d.Get("ptr").(string)
 		ttl := d.Get("ttl").(int)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
-
-		if rec_ptr != dns.Fqdn(rec_ptr) {
-			return fmt.Errorf("Error updating DNS record: \"ptr\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
@@ -174,10 +151,6 @@ func resourceDnsPtrRecordDelete(d *schema.ResourceData, meta interface{}) error 
 
 		rec_name := d.Get("name").(string)
 		rec_zone := d.Get("zone").(string)
-
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error updating DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 

--- a/dns/resource_dns_ptr_record_test.go
+++ b/dns/resource_dns_ptr_record_test.go
@@ -72,10 +72,6 @@ func testAccCheckDnsPtrRecordDestroy(s *terraform.State) error {
 		rec_name := rs.Primary.Attributes["name"]
 		rec_zone := rs.Primary.Attributes["zone"]
 
-		if rec_zone != dns.Fqdn(rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
-
 		rec_fqdn := fmt.Sprintf("%s.%s", rec_name, rec_zone)
 
 		msg := new(dns.Msg)
@@ -104,10 +100,6 @@ func testAccCheckDnsPtrRecordExists(t *testing.T, n string, expected string, rec
 
 		*rec_name = rs.Primary.Attributes["name"]
 		*rec_zone = rs.Primary.Attributes["zone"]
-
-		if *rec_zone != dns.Fqdn(*rec_zone) {
-			return fmt.Errorf("Error reading DNS record: \"zone\" should be an FQDN")
-		}
 
 		rec_fqdn := fmt.Sprintf("%s.%s", *rec_name, *rec_zone)
 

--- a/dns/validators.go
+++ b/dns/validators.go
@@ -1,0 +1,23 @@
+package dns
+
+import (
+	"fmt"
+
+	"github.com/miekg/dns"
+)
+
+func validateZone(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if !dns.IsFqdn(value) {
+		errors = append(errors, fmt.Errorf("DNS zone name %q must be fully qualified: %q", k, value))
+	}
+	return
+}
+
+func validateName(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if dns.IsFqdn(value) {
+		errors = append(errors, fmt.Errorf("DNS record name %q must not be fully qualified: %q", k, value))
+	}
+	return
+}

--- a/dns/validators_test.go
+++ b/dns/validators_test.go
@@ -1,0 +1,49 @@
+package dns
+
+import (
+	"testing"
+)
+
+func TestValidateZone(t *testing.T) {
+	validNames := []string{
+		"example.com.",
+	}
+	for _, v := range validNames {
+		_, errors := validateZone(v, "name")
+		if len(errors) != 0 {
+			t.Fatalf("%q should be a valid DNS zone: %q", v, errors)
+		}
+	}
+
+	invalidNames := []string{
+		"example.com",
+	}
+	for _, v := range invalidNames {
+		_, errors := validateZone(v, "name")
+		if len(errors) == 0 {
+			t.Fatalf("%q should be an invalid DNS zone", v)
+		}
+	}
+}
+
+func TestValidateName(t *testing.T) {
+	validNames := []string{
+		"test",
+	}
+	for _, v := range validNames {
+		_, errors := validateName(v, "name")
+		if len(errors) != 0 {
+			t.Fatalf("%q should be a valid DNS record: %q", v, errors)
+		}
+	}
+
+	invalidNames := []string{
+		"test.",
+	}
+	for _, v := range invalidNames {
+		_, errors := validateName(v, "name")
+		if len(errors) == 0 {
+			t.Fatalf("%q should be an invalid DNS record", v)
+		}
+	}
+}

